index.bs

```diff
@@ -232,27 +232,130 @@ Non-Functional Requirements {#requirements-non-functional}
+  debugging purposes, but may be displayed to the user of the requesting
+  endpoint.
+
+<!-- TODO: Add device type and/or capabilities -->
+
+If a discovering endpoint (acting as a QUIC client) wishes to receive messages
+from an advertising endpoint (acting as a QUIC server) or an advertising
+endpoint (acting as a QUIC server) wishes to send message to a discovering
+endpoint (acting as a client endpoint), it may wish to keep the QUIC connection
+alive.  Once neither side needs to keep the connection alive for the purposes of
+sending or receiving messages, the connection should be closed.  In order to
+keep a QUIC connection alive, an endpoint may send an endpoint-status-request
+message, and any endpoint that receives an endpoint-status-request message
+should send an endpoint-status-response message.  Such messages should be sent
+every 25 seconds.  If, however, other mechanisms or traffic (such as other
+requests and responses) are serving the purpose of keeping the connection alive,
+these messages should be delayed or not used at all.  A QUIC connection should
+not be closed because of idle timeout unless no such mechanisms or traffic has
+been received from the remote side for 35 seconds (allowing for a keep-alive
+frequency of 25 seconds and message delivery delay of 10 seconds).
+
+If a client endpoint wishes to send messages to a server endpoint, the client
+endpoint can connect to the server endpoint "on demand": it does not need to
+keep the connection alive.
+
+The endpoint-info-response message and endpoint-status-response messages may be
+extended to include additional information not defined in this spec.  If done
+adhoc by applications and not in future specs, keys should be chosen to avoid
+collision, such as by choosing large integers or long strings.  
+
+Messages delivery using CBOR and QUIC streams {#control}
+========================================================
+
+Messages are serialized using [CBOR](https://tools.ietf.org/html/rfc7049).  To
+send a group of messages in order, that group of messages must be sent in one
+QUIC stream.  Different groups of messages which are ordered within the group
+but not across groups must be sent in different QUIC streams.  In order to put
+multiple CBOR-serialiezd messages into the the same QUIC stream, the following
+is used.
+
+For each message, the sender must write to the QUIC stream the following: 
+1. A type key representing the type of the message, encoded as a variable-length
+   integer (see Appendix A for type keys) 
+
+2. The message length encoded as a variable-length integer 
+
+3. The messge encoded as CBOR (whose length must match the prefixed length)
+
+Future message type keys may indicate the use of custom message formats that are
+not CBOR, or use of CBOR without the length prefix, so receiving endpoints must
+disregard messages with type keys that it does not recognize and must not assume
+that they are serialized using CBOR nor that they have a length prefix.
```

Comment:
Does this mean then that the endpoint must close the QUIC stream (possibly RST_STREAM + application error)?  If it can't assume there's a length prefix (which makes sense, given we may omit it for some types), it can't skip that message and then continue reading.

If this isn't desirable, one possibility would be to send [length, message] as a CBOR array (and just [message] when there's no length prefix.  In that case, the CBOR parser would still be able to step over it.

Reply:
Ah, good point.  I was mainly thinking of this for messages where there would be only one message in a stream, such as audio or video frames.  But even then, you have a good point about whether the receiving endpoint should reset the stream or not.  

Perhaps I should change this idea and instead require apps to check for capabilities of receiving new message types before sending them.  And we should always tear down the entire connection if an unknown type key is received.

Reply:
OK, I did that: close the connection with a specific error code.